src/rocksdb/src/lib.rs

```diff
@@ -0,0 +1,514 @@
+#![warn(clippy::redundant_closure_call)]
+#![warn(clippy::unnecessary_lazy_evaluations)]
+#![warn(clippy::partialeq_ne_impl)]
+#![warn(clippy::redundant_field_names)]
+#![warn(clippy::transmutes_expressible_as_ptr_casts)]
+#![warn(clippy::unused_async)]
+#![warn(clippy::disallowed_methods)]
+#![warn(clippy::disallowed_macros)]
+#![warn(clippy::disallowed_types)]
+#![warn(clippy::from_over_into)]
+// END LINT CONFIG
+
+//! An async wrapper around RocksDB, designed for the `UPSERT` usecase.
+
+#![warn(missing_docs)]
+
+use std::path::Path;
+use std::time::Instant;
+
+use rocksdb::{
+    DBCompressionType, Env, Error as RocksDBError, Options as RocksDBOptions, WriteOptions, DB,
+};
+use tokio::sync::mpsc::{channel, Sender};
+use tokio::sync::oneshot::{channel as oneshot_channel, Sender as OneshotSender};
+
+use mz_ore::cast::CastLossy;
+use mz_ore::metrics::DeleteOnDropHistogram;
+use mz_persist_types::Codec;
+
+/// An error while creating a Kafka topic.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// An error from the underlying Kafka library.
+    #[error(transparent)]
+    RocksDB(#[from] RocksDBError),
+
+    /// Error when using the instance after RocksDB as errored
+    /// or been shutdown.
+    #[error("rocksDB thread has been shutdown or errored")]
+    RocksDBThreadGoneAway,
+
+    /// Error decoding a value previously written.
+    #[error("failed to decode value: {0}")]
+    DecodeError(String),
+
+    /// A tokio thread used by the implementation panicked.
+    #[error("tokio thread panicked: {0}")]
+    TokioPanic(#[from] tokio::task::JoinError),
+}
+
+/// Options to figure a [`RocksDBInstance`].
+pub struct Options {
+    /// Whether or not to clear state at the instance
+    /// path before starting.
+    pub cleanup_on_new: bool,
+
+    /// Whether or not to write writes
+    /// to the wal.
+    pub use_wal: bool,
+
+    /// Compression type for blocks and blobs.
+    pub compression_type: DBCompressionType,
+
+    /// A possibly shared RocksDB `Env`.
+    pub env: Env,
+}
+
+/// Metrics about an instances usage of RocksDB. User-provided
+/// so the user can choose the labels.
+pub struct RocksDBMetrics {
+    /// Latency in microseonds of multi_gets.
+    pub multi_get_latency_us: DeleteOnDropHistogram<'static, Vec<String>>,
+    /// Size of multi_get batches.
+    pub multi_get_batch_size: DeleteOnDropHistogram<'static, Vec<String>>,
+    /// Latency in microseonds of write batch writes.
+    pub write_latency_us: DeleteOnDropHistogram<'static, Vec<String>>,
+    /// Size of write batches.
+    pub write_batch_size: DeleteOnDropHistogram<'static, Vec<String>>,
+}
+
+impl Options {
+    /// A new `Options` object with reasonable defaults.
+    pub fn new_with_defaults() -> Result<Self, RocksDBError> {
+        Ok(Options {
+            cleanup_on_new: true,
+            use_wal: false,
+            compression_type: DBCompressionType::Snappy,
+            env: rocksdb::Env::new()?,
+        })
+    }
```

Comment:
Does this not want to be `impl Default`?

Reply:
Its fallible